salt/modules/infoblox.py

```diff
@@ -1,508 +1,611 @@
+    # TODO: perhaps cache in __opts__
+    cache_key = 'infoblox_session_{0},{1},{2}'.format(
+        config['api_url'], config['api_user'], config['api_key'])
+    if cache_key in cache:
+        timedelta = int(time.time()) - cache[cache_key]['time']
+        if cache[cache_key]['obj'] and timedelta < 60:
+            return cache[cache_key]['obj']
+    c = {}
+    c['time'] = int(time.time())
+    c['obj'] = libinfoblox.Session(api_sslverify=config['api_sslverify'], api_url=config['api_url'],
+                                   api_user=config['api_user'], api_key=config['api_key'])
+    cache[cache_key] = c
+    return c['obj']
+
+
+def diff_objects(obja, objb):
+    '''
+    Diff two complex infoblox objects.
+    This is used from salt states to detect changes in objects.
+
+    Using `func:nextavailableip` will not cause a diff if the ipaddres is in range
     '''
+    return libinfoblox.diff_obj(obja, objb)
 
-    if infoblox_server is None:
-        infoblox_server = __salt__['pillar.get']('infoblox:server', None)
-    if infoblox_user is None:
-        infoblox_user = __salt__['pillar.get']('infoblox:user', None)
-        log.debug('Infoblox username is "{0}"'.format(infoblox_user))
-    if infoblox_password is None:
-        infoblox_password = __salt__['pillar.get']('infoblox:password', None)
 
-    return infoblox_server, infoblox_user, infoblox_password
+def is_ipaddr_in_ipfunc_range(ipaddr, ipfunc):
+    '''
+    Return true if the ipaddress is in the range of the nextavailableip function
 
+    CLI Example:
 
-def _process_return_data(retData):
+        salt-call infoblox.is_ipaddr_in_ipfunc_range \
+            ipaddr="10.0.2.2" ipfunc="func:nextavailableip:10.0.0.0/8"
     '''
-    generic return processing
+    return libinfoblox.is_ipaddr_in_ipfunc_range(ipaddr, ipfunc)
+
+
+def update_host(name, data, **kwargs):
     '''
-    if retData.status_code == 200:
-        if retData.json():
-            return retData
-        else:
-            log.debug('no data returned from infoblox')
-            return None
-    else:
-        msg = 'Unsuccessful error code {0} returned'.format(retData.status_code)
-        raise CommandExecutionError(msg)
+    Update host record. This is a helper call to update_object.
+
+    Find a hosts `_ref` then call update_object with the record data.
 
+    CLI Example:
 
-def delete_record(name,
-                  dns_view,
-                  record_type,
-                  infoblox_server=None,
-                  infoblox_user=None,
-                  infoblox_password=None,
-                  infoblox_api_version='v1.4.2',
-                  sslVerify=True):
+        salt-call infoblox.update_host name=fqdn data={}
     '''
-    delete a record
+    o = get_host(name=name, **kwargs)
+    return update_object(objref=o['_ref'], data=data, **kwargs)
 
-    name
-        name of the record
 
-    dns_view
-        the DNS view to remove the record from
+def update_object(objref, data, **kwargs):
```

Comment:
ti could be change to something like what the mysql modules has... `def query(database, query, **connection_args)` tho I assume you rather see all the possible arguments spelled out in full.

it would be kinda annoying to add 4 arguments to each function just to pass them along:
```
            api_sslverify: True
            api_url: 'https://INFOBLOX/wapi/v1.2.1'
            api_user: 'username'
            api_key: 'password'
```
or could take in an argument `api_config` that is a dect object... but that is kinda tricky to work with on CLI.

i'm not too motivated to do this change.